File: intellij/src/test/kotlin/dev/nylo/plugin/logs/parse/LogParserTest.kt
package dev.nylo.plugin.logs.parse

import dev.nylo.plugin.logs.NdjsonFixtures
import dev.nylo.plugin.logs.model.NetworkEntry
import dev.nylo.plugin.logs.model.RawLine
import dev.nylo.plugin.logs.model.StandardLine
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.LocalDateTime

class LogParserTest {

    @Test
    fun `parses session fields, log lines and console lines`() {
        val doc = LogParser.parse(NdjsonFixtures.TWO_SESSIONS)
        assertEquals(2, doc.sessions.size)

        val s0 = doc.sessions[0]
        assertEquals("4lfu9t", s0.tag)
        assertEquals("2026-06-17T11-52-55-4lfu9t", s0.fullId)
        assertEquals(LocalDateTime.of(2026, 6, 17, 11, 52, 55), s0.started)
        assertEquals("ios Version 26.5 (Build 23F77)", s0.platform)
        assertEquals("Pretalk 1.0.0", s0.app)
        assertEquals("nylo_framework v7.1.24", s0.version)
        assertEquals("developing", s0.env)
        // The banner block is reconstructed from the session record for display.
        assertTrue(s0.bannerLines.any { it.contains("SESSION  2026-06-17T11-52-55-4lfu9t") })

        val stds = s0.entries.filterIsInstance<StandardLine>()
        assertEquals(3, stds.size)
        assertEquals("debug", stds[0].level)
        assertEquals("4lfu9t", stds[0].sessionTag)
        assertEquals("[AppProvider] setup in 217ms", stds[0].message)
        assertEquals("error", stds[1].level)
        assertEquals("boom", stds[1].message)
        assertTrue(stds[1].stack!!.contains("#0      foo"))
        assertTrue(stds[1].raw.contains("\n#0      foo")) // stack is appended to the displayed line
        // console record -> a StandardLine with no level.
        assertNull(stds[2].level)
        assertEquals("flutter: console output", stds[2].message)
    }

    @Test
    fun `parses network records with kind and structured fields`() {
        val s1 = LogParser.parse(NdjsonFixtures.TWO_SESSIONS).sessions[1]
        assertEquals("9qbhab", s1.tag)

        val nets = s1.entries.filterIsInstance<NetworkEntry>()
        assertEquals(3, nets.size)
        assertEquals(NetworkEntry.Kind.REQUEST, nets[0].kind)
        assertEquals("GET", nets[0].method)
        assertEquals("096232f7", nets[0].requestId)
        assertTrue(nets[0].summary.contains("GET"))
        assertEquals(NetworkEntry.Kind.RESPONSE, nets[1].kind)
        assertEquals(200, nets[1].statusCode)
        assertEquals(NetworkEntry.Kind.ERROR, nets[2].kind)
        assertEquals(500, nets[2].statusCode)
        // Verbose response detail (the body) is folded into the entry's expanded text.
        assertTrue(nets[1].raw.contains("Ann"))
    }

    @Test
    fun `log record without session or level`() {
        val doc = LogParser.parse("""{"t":"log","ts":"2026-06-17T11:52:55","msg":"hello world"}""")
        val line = doc.preamble.filterIsInstance<StandardLine>().single()
        assertNull(line.sessionTag)
        assertNull(line.level)
        assertEquals("hello world", line.message)
        assertEquals(LocalDateTime.of(2026, 6, 17, 11, 52, 55), line.timestamp)
    }

    @Test
    fun `records before the first session are preamble`() {
        val doc = LogParser.parse("""{"t":"console","msg":"boot starting"}""" + "\n" + NdjsonFixtures.TWO_SESSIONS)
        assertEquals(2, doc.sessions.size)
        assertTrue(doc.preamble.any { it.raw.contains("boot starting") })
    }

    @Test
    fun `non-json and unknown records become raw lines`() {
        val doc = LogParser.parse("not json at all\n" + """{"t":"mystery","msg":"x"}""")
        val raws = doc.preamble.filterIsInstance<RawLine>()
        assertTrue(raws.any { it.raw == "not json at all" })
        assertTrue(raws.any { it.raw.contains("mystery") })
    }

    @Test
    fun `carriage returns are normalized out`() {
        val one = """{"t":"console","msg":"line one"}"""
        val two = """{"t":"console","msg":"line two"}"""
        val doc = LogParser.parse(one + "\r\n" + two + "\r\n")
        assertEquals(listOf("line one", "line two"), doc.preamble.filterIsInstance<StandardLine>().map { it.message })
        assertTrue(doc.preamble.all { '\r' !in it.raw })
    }

    @Test
    fun `json-escaped carriage returns inside fields never reach the raw text`() {
        // A \r inside a JSON string survives whole-file normalization (it exists in the file only
        // as the two-char escape); a literal CR in `raw` would crash the editor Document.
        val record = """{"t":"log","ts":"2026-06-26 10:00:00","session":"s1","level":"error",""" +
            """"msg":"line1\r\nline2","stack":"#0 a\r#1 b"}"""
        val line = LogParser.parse(record).preamble.single() as StandardLine
        assertTrue('\r' !in line.raw)
        assertTrue(line.raw.contains("line1\nline2"))
        assertTrue(line.stack!!.contains("#0 a\n#1 b"))
    }

    @Test
    fun `parses iso timestamps with negative utc offsets`() {
        val doc = LogParser.parse("""{"t":"log","ts":"2026-06-26T10:32:59-05:00","msg":"hi"}""")
        val line = doc.preamble.single() as StandardLine
        assertEquals(LocalDateTime.of(2026, 6, 26, 10, 32, 59), line.timestamp)
    }
}

File: vscode/src/core/logs/logParser.test.ts
import * as assert from 'assert';
import { TWO_SESSIONS } from './logParser.fixtures';
import { parseLog } from './logParser';
import { NetworkEntry, RawLine, StandardLine } from './logModel';

describe('logParser', () => {
  it('parses session fields, log lines and console lines', () => {
    const doc = parseLog(TWO_SESSIONS);
    assert.strictEqual(doc.sessions.length, 2);

    const s0 = doc.sessions[0];
    assert.strictEqual(s0.tag, '4lfu9t');
    assert.strictEqual(s0.fullId, '2026-06-17T11-52-55-4lfu9t');
    assert.strictEqual(s0.started, Date.UTC(2026, 5, 17, 11, 52, 55));
    assert.strictEqual(s0.platform, 'ios Version 26.5 (Build 23F77)');
    assert.strictEqual(s0.app, 'Pretalk 1.0.0');
    assert.strictEqual(s0.version, 'nylo_framework v7.1.24');
    assert.strictEqual(s0.env, 'developing');
    assert.ok(s0.bannerLines.some((l) => l.includes('SESSION  2026-06-17T11-52-55-4lfu9t')));

    const stds = s0.entries.filter((e): e is StandardLine => e.entryType === 'standard');
    assert.strictEqual(stds.length, 3);
    assert.strictEqual(stds[0].level, 'debug');
    assert.strictEqual(stds[0].sessionTag, '4lfu9t');
    assert.strictEqual(stds[0].message, '[AppProvider] setup in 217ms');
    assert.strictEqual(stds[1].level, 'error');
    assert.strictEqual(stds[1].message, 'boom');
    assert.ok(stds[1].stack!.includes('#0      foo'));
    assert.ok(stds[1].raw.includes('\n#0      foo'));
    assert.strictEqual(stds[2].level, null);
    assert.strictEqual(stds[2].message, 'flutter: console output');
  });

  it('parses network records with kind and structured fields', () => {
    const s1 = parseLog(TWO_SESSIONS).sessions[1];
    assert.strictEqual(s1.tag, '9qbhab');

    const nets = s1.entries.filter((e): e is NetworkEntry => e.entryType === 'network');
    assert.strictEqual(nets.length, 3);
    assert.strictEqual(nets[0].netKind, 'request');
    assert.strictEqual(nets[0].method, 'GET');
    assert.strictEqual(nets[0].requestId, '096232f7');
    assert.ok(nets[0].summary.includes('GET'));
    assert.strictEqual(nets[1].netKind, 'response');
    assert.strictEqual(nets[1].statusCode, 200);
    assert.strictEqual(nets[2].netKind, 'error');
    assert.strictEqual(nets[2].statusCode, 500);
    assert.ok(nets[1].raw.includes('Ann'));
  });

  it('log record without session or level', () => {
    const doc = parseLog('{"t":"log","ts":"2026-06-17T11:52:55","msg":"hello world"}');
    const line = doc.preamble.filter((e): e is StandardLine => e.entryType === 'standard')[0];
    assert.strictEqual(line.sessionTag, null);
    assert.strictEqual(line.level, null);
    assert.strictEqual(line.message, 'hello world');
    assert.strictEqual(line.timestamp, Date.UTC(2026, 5, 17, 11, 52, 55));
  });

  it('records before the first session are preamble', () => {
    const doc = parseLog('{"t":"console","msg":"boot starting"}\n' + TWO_SESSIONS);
    assert.strictEqual(doc.sessions.length, 2);
    assert.ok(doc.preamble.some((e) => e.raw.includes('boot starting')));
  });

  it('non-json and unknown records become raw lines', () => {
    const doc = parseLog('not json at all\n' + '{"t":"mystery","msg":"x"}');
    const raws = doc.preamble.filter((e): e is RawLine => e.entryType === 'raw');
    assert.ok(raws.some((r) => r.raw === 'not json at all'));
    assert.ok(raws.some((r) => r.raw.includes('mystery')));
  });

  it('carriage returns are normalized out', () => {
    const one = '{"t":"console","msg":"line one"}';
    const two = '{"t":"console","msg":"line two"}';
    const doc = parseLog(one + '\r\n' + two + '\r\n');
    assert.deepStrictEqual(
      doc.preamble.filter((e): e is StandardLine => e.entryType === 'standard').map((e) => e.message),
      ['line one', 'line two'],
    );
    assert.ok(doc.preamble.every((e) => !e.raw.includes('\r')));
  });

  it('json-escaped carriage returns inside fields never reach the raw text', () => {
    // A \r inside a JSON string survives whole-file normalization (it exists in the file only
    // as the two-char escape), so the field accessor must normalize it out of `raw`.
    const record =
      '{"t":"log","ts":"2026-06-26 10:00:00","session":"s1","level":"error",' +
      '"msg":"line1\\r\\nline2","stack":"#0 a\\r#1 b"}';
    const line = parseLog(record).preamble[0] as StandardLine;
    assert.ok(!line.raw.includes('\r'));
    assert.ok(line.raw.includes('line1\nline2'));
    assert.ok(line.stack!.includes('#0 a\n#1 b'));
  });

  it('parses iso timestamps with negative utc offsets', () => {
    const doc = parseLog('{"t":"log","ts":"2026-06-26T10:32:59-05:00","msg":"hi"}');
    const line = doc.preamble[0] as StandardLine;
    assert.strictEqual(line.timestamp, Date.UTC(2026, 5, 26, 10, 32, 59));
  });
});
